File: sdk/agent-core-ts/src/agent.ts
import { AgentAction, AgentState } from './types.js'
import type { Strategy, MutableAgentState } from './strategy.js'
import { AgentRole } from './role.js'

export class Agent {
  readonly id: string
  private _strategy: Strategy
  private _running = false
  private _rpcEndpoint = 'https://api.devnet.solana.com'
  private _network = 'devnet'
  private _actions: AgentAction[] = []
  private _abortController: AbortController | null = null
  role: AgentRole = AgentRole.Worker

  constructor(id: string, strategy: Strategy) {
    this.id = id
    this._strategy = strategy
  }

  get isRunning(): boolean { return this._running }
  get strategy(): Strategy { return this._strategy }

  setStrategy(strategy: Strategy): void {
    this._strategy = strategy
  }

  setRpc(url: string): void {
    this._rpcEndpoint = url
    const url_ = url.toLowerCase()
    if (url_.includes('devnet')) this._network = 'devnet'
    else if (url_.includes('testnet')) this._network = 'testnet'
    else if (url_.includes('mainnet')) this._network = 'mainnet-beta'
  }

  recordAction(actionType: string, details: string, txSignature?: string, slot?: number): void {
    this._actions.push({
      timestamp: new Date().toISOString(),
      action_type: actionType,
      details,
      tx_signature: txSignature ?? null,
      slot: slot ?? null,
      latency_ms: 0,
    })
    // cap at 500
    if (this._actions.length > 500) this._actions.splice(0, this._actions.length - 500)
  }

  state(): AgentState {
    return {
      is_running: this._running,
      actions: [...this._actions],
      rpc_endpoint: this._rpcEndpoint,
      network: this._network,
      strategy: this._strategy.name,
    }
  }

  // Build the MutableAgentState view used by strategies.
  private makeMutable(): MutableAgentState {
    const agent = this
    return {
      get id() { return agent.id },
      get rpcEndpoint() { return agent._rpcEndpoint },
      get network() { return agent._network },
      recordAction: this.recordAction.bind(this),
      snapshot: this.state.bind(this),
    }
  }

  // Deliver a message to this agent's strategy and return its response.
  // Mirrors Rust agent.get_strategy().handle_message(text, state_arc).
  async handleMessage(text: string): Promise<string> {
    return this._strategy.handleMessage(text, this.makeMutable())
  }

  async start(): Promise<boolean> {
    if (this._running) return false
    this._running = true
    this._abortController = new AbortController()
    const signal = this._abortController.signal
    const mutable = this.makeMutable()

    this._strategy.run(mutable, signal)
      .catch((e) => {
        if (!signal.aborted) {
          this.recordAction('strategy-error', String(e))
        }
      })
      .finally(() => {
        this._running = false
      })

    this.recordAction('strategy-start', `started ${this._strategy.name}`)
    return true
  }

  stop(): boolean {
    if (!this._running) return false
    this._abortController?.abort()
    this._running = false
    return true
  }
}

File: desktop/agent-core/src/agent.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::sync::{Arc, Mutex};

use crate::strategy::{RpcPollStrategy, Strategy};

/// A single recorded action taken by an agent.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct AgentAction {
    pub timestamp: DateTime<Utc>,
    /// Short machine-readable label (e.g. `"rpc-poll"`, `"url-generated"`).
    pub action_type: String,
    /// Human-readable description of what happened.
    pub details: String,
    pub tx_signature: Option<String>,
    pub slot: Option<u64>,
    /// Wall-clock latency of the operation in milliseconds.
    pub latency_ms: u64,
}

/// Serialisable snapshot of an agent's current state, suitable for IPC.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct AgentState {
    pub is_running: bool,
    pub actions: Vec<AgentAction>,
    pub rpc_endpoint: String,
    pub network: String,
    /// Name of the active [`Strategy`] implementation (e.g. `"rpc-poll"`).
    pub strategy: String,
}

impl Default for AgentState {
    fn default() -> Self {
        Self {
            is_running: false,
            actions: Vec::new(),
            rpc_endpoint: "https://api.devnet.solana.com".to_string(),
            network: "devnet".to_string(),
            strategy: "rpc-poll".to_string(),
        }
    }
}

/// A single autonomous agent with a pluggable [`Strategy`] and an action log.
///
/// All mutable state is protected by `Mutex` so the agent can be shared
/// across threads (required by the Tauri `State` extractor and `tokio::spawn`).
pub struct Agent {
    state: Arc<Mutex<AgentState>>,
    strategy: Mutex<Arc<dyn Strategy>>,
}

impl Default for Agent {
    fn default() -> Self {
        Self::new()
    }
}

impl Agent {
    /// Create an agent with the default [`RpcPollStrategy`].
    pub fn new() -> Self {
        Self::with_strategy(Arc::new(RpcPollStrategy::new()))
    }

    /// Create an agent using a specific strategy. The strategy name is recorded
    /// in the initial [`AgentState`].
    pub fn with_strategy(strategy: Arc<dyn Strategy>) -> Self {
        let mut state = AgentState::default();
        state.strategy = strategy.name().to_string();
        Self {
            state: Arc::new(Mutex::new(state)),
            strategy: Mutex::new(strategy),
        }
    }

    /// Return a cloned snapshot of the current state.
    pub fn state(&self) -> AgentState {
        self.state.lock().expect("agent state lock poisoned").clone()
    }

    /// Override the RPC endpoint (takes effect on the next strategy tick).
    pub fn set_rpc(&self, url: String) {
        self.state.lock().expect("agent state lock poisoned").rpc_endpoint = url;
    }

    /// Override the network label (e.g. `"mainnet"`, `"devnet"`).
    pub fn set_network(&self, network: String) {
        self.state.lock().expect("agent state lock poisoned").network = network;
    }

    /// Configure the agent's RPC endpoint and network from a Triton config.
    pub fn set_triton(&self, config: &crate::triton::TritonConfig) {
        let mut s = self.state.lock().expect("agent state lock poisoned");
        s.rpc_endpoint = config.rpc_url.clone();
        s.network = config.network.clone();
    }

    /// Append an action to the agent's log.
    pub fn record_action(&self, action: AgentAction) {
        self.state
            .lock()
            .expect("agent state lock poisoned")
            .actions
            .push(action);
    }

    /// Hot-swap the running strategy. The new strategy name is written to state
    /// immediately; the old strategy loop exits on its next `is_running` check.
    pub fn set_strategy(&self, strategy: Arc<dyn Strategy>) {
        let mut s = self.state.lock().expect("agent state lock poisoned");
        s.strategy = strategy.name().to_string();
        *self.strategy.lock().expect("strategy lock poisoned") = strategy;
    }

    /// Mark the agent as running and spawn the strategy loop on the Tokio runtime.
    pub async fn start_monitoring(&self) -> anyhow::Result<()> {
        self.state.lock().expect("agent state lock poisoned").is_running = true;

        let strategy = Arc::clone(&*self.strategy.lock().expect("strategy lock poisoned"));
        let state = Arc::clone(&self.state);
        tokio::spawn(async move {
            strategy.run(state).await;
        });

        Ok(())
    }

    /// Signal the strategy loop to exit by setting `is_running = false`.
    pub fn stop(&self) {
        self.state.lock().expect("agent state lock poisoned").is_running = false;
    }

    /// Return the live strategy so callers can dispatch messages to it.
    pub fn get_strategy(&self) -> Arc<dyn Strategy> {
        Arc::clone(&*self.strategy.lock().expect("strategy lock poisoned"))
    }

    /// Return the shared state Arc so strategy.handle_message can mutate it.
    pub fn state_arc(&self) -> Arc<Mutex<AgentState>> {
        Arc::clone(&self.state)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_agent_state() {
        let agent = Agent::new();
        assert!(!agent.state().is_running);
    }

    #[tokio::test]
    async fn test_monitoring_starts() {
        let agent = Agent::new();
        agent.start_monitoring().await.unwrap();
        assert!(agent.state().is_running);
        agent.stop();
    }

    #[test]
    fn test_transfer_url_roundtrip() {
        let fields = crate::solana_pay::TransferUrlFields {
            recipient: "7xKF3rO1jW...".to_string(),
            amount: Some(0.01),
            spl_token: None,
            reference: None,
            label: Some("Demo".to_string()),
            message: Some("Thanks!".to_string()),
            memo: None,
        };
        let url = crate::solana_pay::encode_transfer_url(&fields);
        let parsed = crate::solana_pay::parse_url(&url).unwrap();
        match parsed {
            crate::solana_pay::ParsedUrl::Transfer(t) => {
                assert_eq!(t.recipient, fields.recipient);
                assert_eq!(t.amount, fields.amount);
                assert_eq!(t.label, fields.label);
                assert_eq!(t.message, fields.message);
            }
            _ => panic!("Expected transfer URL"),
        }
    }
}
